diff -Nur openresty-1.15.8.3/bundle/lua-resty-core-0.1.17/lib/ngx/ssl.lua openresty-1.15.8.3.patched/bundle/lua-resty-core-0.1.17/lib/ngx/ssl.lua
--- openresty-1.15.8.3/bundle/lua-resty-core-0.1.17/lib/ngx/ssl.lua	2019-08-12 14:08:39.000000000 +0800
+++ openresty-1.15.8.3.patched/bundle/lua-resty-core-0.1.17/lib/ngx/ssl.lua	2019-08-12 14:07:33.000000000 +0800
@@ -25,6 +25,7 @@
 local ngx_lua_ffi_ssl_set_der_private_key
 local ngx_lua_ffi_ssl_raw_server_addr
 local ngx_lua_ffi_ssl_server_name
+local ngx_lua_ffi_ssl_server_port
 local ngx_lua_ffi_ssl_raw_client_addr
 local ngx_lua_ffi_cert_pem_to_der
 local ngx_lua_ffi_priv_key_pem_to_der
@@ -53,6 +54,9 @@
     int ngx_http_lua_ffi_ssl_server_name(ngx_http_request_t *r, char **name,
         size_t *namelen, char **err);
 
+    int ngx_http_lua_ffi_ssl_server_port(ngx_http_request_t *r,
+        int *port, char **err);
+
     int ngx_http_lua_ffi_ssl_raw_client_addr(ngx_http_request_t *r, char **addr,
         size_t *addrlen, int *addrtype, char **err);
 
@@ -87,6 +91,7 @@
         C.ngx_http_lua_ffi_ssl_set_der_private_key
     ngx_lua_ffi_ssl_raw_server_addr = C.ngx_http_lua_ffi_ssl_raw_server_addr
     ngx_lua_ffi_ssl_server_name = C.ngx_http_lua_ffi_ssl_server_name
+    ngx_lua_ffi_ssl_server_port = C.ngx_http_lua_ffi_ssl_server_port
     ngx_lua_ffi_ssl_raw_client_addr = C.ngx_http_lua_ffi_ssl_raw_client_addr
     ngx_lua_ffi_cert_pem_to_der = C.ngx_http_lua_ffi_cert_pem_to_der
     ngx_lua_ffi_priv_key_pem_to_der = C.ngx_http_lua_ffi_priv_key_pem_to_der
@@ -263,6 +268,27 @@
 end
 
 
+function _M.server_port()
+    local r = get_request()
+    if not r then
+        error("no request found")
+    end
+
+    local sizep = get_size_ptr()
+
+    local rc = ngx_lua_ffi_ssl_server_port(r, intp, errmsg)
+    if rc == FFI_OK then
+        return intp[0]
+    end
+
+    if rc == FFI_DECLINED then
+        return nil
+    end
+
+    return nil, ffi_str(errmsg[0])
+end
+
+
 function _M.raw_client_addr()
     local r = get_request()
     if not r then
diff -Nur openresty-1.15.8.3/bundle/ngx_lua-0.10.15/src/ngx_http_lua_ssl_certby.c openresty-1.15.8.3.patched/bundle/ngx_lua-0.10.15/src/ngx_http_lua_ssl_certby.c
--- openresty-1.15.8.3/bundle/ngx_lua-0.10.15/src/ngx_http_lua_ssl_certby.c	2019-08-12 14:08:39.000000000 +0800
+++ openresty-1.15.8.3.patched/bundle/ngx_lua-0.10.15/src/ngx_http_lua_ssl_certby.c	2019-08-12 14:07:33.000000000 +0800
@@ -882,6 +882,39 @@
 
 
 int
+ngx_http_lua_ffi_ssl_server_port(ngx_http_request_t *r,
+    int *port, char **err)
+{
+    ngx_ssl_conn_t       *ssl_conn;
+    ngx_connection_t     *c;
+    int                  p;
+    if (r->connection == NULL || r->connection->ssl == NULL) {
+        *err = "bad request";
+        return NGX_ERROR;
+    }
+    ssl_conn = r->connection->ssl->connection;
+    if (ssl_conn == NULL) {
+        *err = "bad ssl conn";
+        return NGX_ERROR;
+    }
+
+    c = ngx_ssl_get_connection(ssl_conn);
+
+    if (ngx_connection_local_sockaddr(c, NULL, 0) != NGX_OK) {
+        return 0;
+    }
+
+    p = ngx_inet_get_port(c->local_sockaddr);
+
+    if (p > 0 && p < 65536) {
+        *port = p;
+    }
+
+    return NGX_OK;
+}
+
+
+int
 ngx_http_lua_ffi_ssl_raw_client_addr(ngx_http_request_t *r, char **addr,
     size_t *addrlen, int *addrtype, char **err)
 {
